executor/src/witgen/machines/block_machine.rs

```diff
@@ -315,167 +319,83 @@ impl<'a, T: FieldElement> BlockMachine<'a, T> {
-        fixed_data: &'a FixedData<T>,
-        left: &[AffineExpression<&'a PolynomialReference, T>],
-        right: &'a SelectedExpressions<T>,
-        identity: IdentityInSequence,
-        identity_processor: &mut IdentityProcessor<'a, '_, T>,
-    ) -> EvalResult<'a, T> {
-        match identity {
-            IdentityInSequence::Internal(i) => identity_processor.process_identity(
-                self.identities[i],
-                &self.get_current_row_pair(row, fixed_data),
-            ),
-            IdentityInSequence::OuterQuery => {
-                assert!(row as usize % self.block_size == self.block_size - 1);
-                identity_processor.process_link(
-                    left,
-                    right,
-                    &self.get_current_row_pair(row, fixed_data),
-                )
-            }
-        }
-    }
-
-    fn apply_updates(
-        &mut self,
-        row: DegreeType,
-        identity: IdentityInSequence,
-        updates: &EvalValue<&'a PolynomialReference, T>,
-        left_mut: &mut [AffineExpression<&'a PolynomialReference, T>],
-    ) -> bool {
-        if updates.constraints.is_empty() {
-            return false;
+    /// Takes a block of rows, which contains the last row of the previous block
+    /// and the first row of the next block. The first row of the next block is ignored,
+    /// the last row of the previous block is merged with the first row of the next block.
+    /// This is necessary to handle non-rectangular block machines, which already use
+    /// unused cells in the previous block.
+    fn append_block(&mut self, mut new_block: Vec<Row<'a, T>>) -> Result<(), EvalError<T>> {
+        if self.rows() + self.block_size as DegreeType >= self.fixed_data.degree {
+            return Err(EvalError::RowsExhausted);
         }
 
-        match identity {
-            IdentityInSequence::Internal(index) => {
-                log::trace!("    Updates from: {}", self.identities[index])
-            }
-            IdentityInSequence::OuterQuery => log::trace!("    Updates from: outer query"),
-        };
-
-        let (before, after) = self.data.split_at_mut(row as usize + 1);
-        let current = before.last_mut().unwrap();
-        let next = after.first_mut().unwrap();
-
-        let mut row_updater = RowUpdater::new(current, next, row);
-        for (poly, c) in &updates.constraints {
-            if self.witness_cols.contains(&poly.poly_id()) {
-                row_updater.apply_update(poly, c);
-            } else if let Constraint::Assignment(v) = c {
-                for l in left_mut.iter_mut() {
-                    log::trace!("      => {} (outer) = {}", poly, v);
-                    l.assign(poly, *v);
+        assert_eq!(new_block.len(), self.block_size + 2);
+
+        // 1. Ignore the first row of the next block:
+        new_block.pop();
+        // 2. Take out the last row from the previous block
+        let (mut last_row, new_block): (Vec<_>, Vec<_>) =
```

Comment:
Maybe use `slice::split_first_mut`?

Reply:
Hmm, this doesn't quite what I need here, because it won't let me move `new_block`, as it's just a slice.
I changed the code anyway, now I use `new_block.get_mut(0).unwrap()`, mutate the row, remove the existing row and then append `new_block`.